app/controllers/DataUploadController.scala

```diff
@@ -116,10 +128,60 @@ class DataUploadController @Inject()(sessionService: SessionService,
       )
   }
 
-  def process(res: List[UpscanCallback], empRef: String)
-             (hc: HeaderCarrier, schemeInfo: SchemeInfo, request: Request[_]): List[Future[(Int, Int)]] = {
-    for {
-      callbackData <- res
-    } yield fileProcessService.processCsvFile(callbackData, empRef)(hc, schemeInfo,request)
+  def processCsvFileDataFromFrontendV2(empRef:String): Action[JsValue] = authorisedActionWithBody(empRef) {
+    implicit request: Request[JsValue] =>
+      Logger.warn("receiving request on V2 processCsvFileDataFromFrontend")
+      val startTime =  System.currentTimeMillis()
+      request.body.validate[UpscanCsvFileData].fold(
+        valid = res => {
+          val schemeInfo: SchemeInfo = res.schemeInfo
+          Logger.debug("SCHEME TYPE: " + schemeInfo.schemeType)
+          deliverFileProcessingMetrics(startTime)
+
+          val processedFiles: List[Future[Either[Throwable, CsvFileSubmissions]]] = processCsvService.processFilesNew(res, streamFile)
+
+          val extractedSchemeData: Seq[Future[Either[Throwable, (Int, Int)]]] = processedFiles.map{ submission =>
+            submission.flatMap(processCsvService.extractSchemeDataNew(res.schemeInfo, empRef, _))}
+
+          Future.sequence(extractedSchemeData).flatMap{ oneFileResults => oneFileResults.find(_.isLeft) match {
+            case Some(Left(throwable: ERSFileProcessingException)) =>
+              Logger.error(s"[DataUploadController][processCsvFileDataFromFrontend] ERS file processing exception: ${throwable.message}")
+              deliverFileProcessingMetrics(startTime)
+              Future(Accepted(throwable.message))
+            case Some(Left(throwable)) =>
+              Logger.error(s"[DataUploadController][processCsvFileDataFromFrontend] Unknown exception: ${throwable.getMessage}, full exception: $throwable")
+              deliverFileProcessingMetrics(startTime)
+              Future(InternalServerError)
+            case None =>
+              val result: Seq[(Int, Int)] = oneFileResults.map(_.right.get)
+              val totalRowCount = result.foldLeft(0) ((accum,inputTuple) => accum + inputTuple._2)
```

Comment:
Following on from our conversation... if we know that this Seq[(Int, Int)] is only going to have a length of one (since we're already dealing with a single file), should we not just take noOfRows directly from our sequence entry?

Same goes for the sum of slices further down.

Reply:
the sequence can very much have a length of more than 1 (if the file has more than 10,000 rows), iirc each of the entries in the sequence represents a 'chunk'

Reply:
after the discussion: very good spot, this sequence does indeed only have one entry for each file, but the variable name was confusing (I'll change it): `oneFileResults` is actually a sequence containing the 'result' for each of the files being uploaded, thus totalRowCount is, in fact, the row count across **all** files being uploaded.

I agree that this is a bit funky, but I did some testing and found out that that's the functionality of ers-file-validator from before our change, so I'd be hesitant about changing it with this ticket - files still seem to be uploaded correctly, but I'm not sure why it's decided to store the total row count rather than row count per file.